src/charts/tau.gpl.js

```diff
@@ -0,0 +1,241 @@
+
+        var domX = domainX.length === 0 ? [null] : domainX;
+        var domY = domainY.length === 0 ? [null] : domainY;
+
+        return _(domY).reduce(
+            (memo, rowVal) => {
+
+                return memo.concat(_(domX).map((colVal) => {
+
+                    var r = {};
+
+                    if (dimX) {
+                        r[dimX] = convert(colVal);
+                    }
+
+                    if (dimY) {
+                        r[dimY] = convert(rowVal);
+                    }
+
+                    return r;
+                }));
+            },
+            []);
+    },
+
+    'none': function(datus, dimX, dimY, pipe) {
+        return [null];
+    }
+};
+
+var calcBaseFrame = (unitExpression, baseFrame) => {
+
+    var tmpFrame = _.pick(baseFrame || {}, 'source', 'pipe');
+
+    var srcAlias = unitExpression.source;
+    var bInherit = unitExpression.inherit;
+    var ownFrame = {source: srcAlias, pipe: []};
+
+    if (bInherit && (ownFrame.source !== tmpFrame.source)) {
+        throw new Error(`base [${tmpFrame.source}] and own [${ownFrame.source}] sources should be equal to apply inheritance`);
+    }
+
+    return bInherit ? tmpFrame : ownFrame;
+};
+
+export class GPL extends Emitter {
+
+    constructor(config) {
+
+        super();
+
+        this._svg = null;
+        this._filtersStore = {filters: {}, tick: 0};
+        this._layout = getLayout();
+
+        this._initialize(config);
+    }
+
+    _initialize(config) {
+
+        this.config = config;
+
+        this.unitSet = UnitsRegistry;
+
+        this.sources = config.sources;
+
+        this.scalesCreator = new ScalesFactory(config.sources);
+
+        this.scales = config.scales;
+
+        this.trans = config.trans;
+    }
+
+    render(target, xSize) {
+
+        var targetNode = d3.select(target).node();
+        if (targetNode === null) {
+            throw new Error('Target element not found');
+        }
+
+        targetNode.appendChild(this._layout.layout);
+
+        var containerNode = this._layout.content;
+        var container = d3.select(this._layout.content);
+        containerNode.innerHTML = '';
+
+        var size = _.clone(xSize) || {};
+        if (!size.width || !size.height) {
+            // size = _.defaults(size, utilsDom.getContainerSize(containerNode.parentNode));
+            size = _.defaults(size, utilsDom.getContainerSize(targetNode));
+        }
+
+        // expand units structure
+        this.root = this.expandUnitsStructure(this.config.unit);
+
+        this.root.options = {
+            container: container.append('svg').attr(_.extend({'class': `${CSS_PREFIX}svg`}, size)),
+            left     : 0,
+            top      : 0,
+            width    : size.width,
+            height   : size.height
+        };
+
+        this.drawUnitsStructure(this.root);
+    }
+
+    expandUnitsStructure(rootUnit) {
+
+        var buildRecursively = (root, parentPipe) => {
```

Comment:
It reduntant function you can combine this funciton with expandUnitsStructure,
like this
`expandUnitsStructure(rootUnit, parentPipe = [])`
or
make this function pure and move to utils